Fix part of #13822: Added beam jobs for topic validation

Lawful2002 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #13822 .
2. This PR does the following: Added beam job for validating topic model

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
<img width="527" alt="image" src="https://user-images.githubusercontent.com/73544247/158325268-a0837245-ba4a-4bb5-9b42-6f6b6f0c04f4.png">


<!--
Add videos/screenshots of the user-facing interface in various display sizes (mainly phone, tablet, and desktop display size) to demonstrate that the changes made in this PR work correctly. Please also include videos/screenshots of the developer tools browser console, so that we can be sure that there are no console errors.
If the PR changes the UI in any of the files listed in rtl_css.py (i.e, those that have a separate .rtl.css file for styling), make sure to add screenshots with the site language set to Arabic as well.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

Lawful2002 committed:
added beam job

Lawful2002 committed:
added tests

Lawful2002 committed:
added backend tests

Lawful2002 committed:
fixed typo

vojtechjelinek reviewed (changes requested):
Looks good! After @lkbhitesh07 reviews this feel free to submit testing request.

lkbhitesh07 reviewed (approved):
Everything looks great, Thank you so much.

U8NWXD commented:
@Lawful2002 I need to check that the backend test shards are running within the expected time, and to do so I need the backend tests to be passing. Can you please fix the backend tests and then reassign me?

Lawful2002 commented:
@U8NWXD It seems like core/domain/subscription_services.py is not completely covered. I'm not sure if it is related to this PR.

U8NWXD commented:
@Lawful2002 we have full backend test coverage on develop, so I would assume by default that any incomplete coverage is due to this PR. While backend flakes are rare, they do happen occasionally, and I don't see how your code changes would affect coverage of subscription_services.py. To check whether this is a flake, could you run the backend tests locally to see if you can reproduce this failure?

Lawful2002 commented:
@U8NWXD Apparently re-running the job fixed it. PTAL. Thanks!

vojtechjelinek commented:
Hey, this fails on the backup server with this error

 ```
File "core/jobs/batch_jobs/topic_validation_jobs.py", line 54, 
  in <lambda> beam.Filter(lambda topic: len(topic[1]) > 39) 
TypeError: object of type 'NoneType' has no len() [while running 'Filter topics with abb. name greater than 39-ptransform-110'] 
```

Lawful2002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into topic-beam-jobs

Lawful2002 committed:
added beam jobs for page_title_fragment_for_web

Lawful2002 committed:
removed unnecessary job

Lawful2002 commented:
@vojtechjelinek, after discussion with Sean, it has been decided that we won't be validating the `abbreviated_name` field (it will be eventually removed). I have removed the job for the same. Also I have added the job for validating `page_title_fragment_for_web` in this PR. PTAL

vojtechjelinek commented:
@lkbhitesh07 Please take a pass through the new changes.

lkbhitesh07 commented on core/jobs/batch_jobs/topic_validation_jobs.py:
```diff
@@ -0,0 +1,89 @@
+                beam.Filter(lambda topic: (
+                    len(topic[1]) > 50 or
+                    len(topic[1]) < 5
+                    ))
```
Is there any possibility that `page_title_fragment_for_web` might be None? I think we should also check for the None values otherwise it will throw an error in the future.

Lawful2002 replied:
Done!

lkbhitesh07 commented on core/jobs/batch_jobs/topic_validation_jobs_test.py:
```diff
@@ -0,0 +1,122 @@
+    def setUp(self):
+        super().setUp()
+
+        long_fragment = 'abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxy'
```
```suggestion
        LONG_FRAGMENT = 'abcdefghijklmnopqrstuvwxyzabcdefghijklmnopqrstuvwxy'
```

Lawful2002 replied:
Done!

lkbhitesh07 commented on core/jobs/batch_jobs/topic_validation_jobs_test.py:
```diff
@@ -0,0 +1,122 @@
+            language_code=constants.DEFAULT_LANGUAGE_CODE,
+            url_fragment='topic_one',
+            practice_tab_is_displayed=True,
+            page_title_fragment_for_web=long_fragment
```
```suggestion
            page_title_fragment_for_web=LONG_FRAGMENT
```

Lawful2002 replied:
Done!

lkbhitesh07 commented on core/jobs/batch_jobs/topic_validation_jobs_test.py:
```diff
@@ -0,0 +1,122 @@
+            practice_tab_is_displayed=True,
+            page_title_fragment_for_web=long_fragment
+        )
+
```
If we are planning to handle the case for `None` you can add a test.

Lawful2002 replied:
Done!

lkbhitesh07 reviewed (changes requested):
Left some comments.
Thanks

Lawful2002 committed:
addressed comments

lkbhitesh07 commented on core/jobs/batch_jobs/topic_validation_jobs.py:
```diff
@@ -0,0 +1,92 @@
+
+        topic_ids_with_exceeding_page_title_fragment = (
+            total_topics
+            | 'Filter topics with title frag greater than 50 or less than 5' >>
```
```suggestion
            | 'Filter topics with title frag length greater than 50 or less than 5' >>
```

Lawful2002 replied:
Done!

lkbhitesh07 reviewed (approved):
Just a small nit, else everything looks good apart from failing tests.
Thanks

Lawful2002 committed:
Update core/jobs/batch_jobs/topic_validation_jobs.py

Co-authored-by: Hitesh Tomar <44300735+lkbhitesh07@users.noreply.github.com>

Lawful2002 committed:
fixed linting

vojtechjelinek commented:
@Lawful2002 Can you please submit a new testing request?

vojtechjelinek commented:
@Lawful2002 Hey, all topics have correct title fragment, the output of the job is "TOPICS SUCCESS: 9"

Lawful2002 commented:
Thanks! @vojtechjelinek. I will start working on adding backend validation.

Lawful2002 committed:
added check for min length

Lawful2002 committed:
Merge branch 'web-frag-backend-val' of https://github.com/Lawful2002/oppia into topic-beam-jobs

Lawful2002 committed:
removed jobs

Lawful2002 committed:
fixed test

Lawful2002 committed:
fixed linting

Lawful2002 committed:
fixed tests

Lawful2002 commented:
The frontend validation already exists for this field. However, as the `page_title_fragment_for_web` needs to be non-empty (between 5 to 50 characters inclusive), adding the backend validation in the current state seems to be tricky as this field is not initially filled in the topic creation modal.

There are two alternatives, that I could think of:
1. Add a 'default' value for this field (this could be problematic, as the user may not even change the default at all)
2. Make this a 'required' field (which it isn't at the moment), and have the user enter it in the initial modal that appears.

@vojtechjelinek What do you think would be the best way to tackle this?

<img width="315" alt="image" src="https://user-images.githubusercontent.com/73544247/161492775-7294c722-07bd-4d2e-8365-0dac5d3b8ef5.png">

seanlip commented:
Go ahead and do (2). I think that is fine. If you like, you can do the same thing for the meta tag as well.

Lawful2002 committed:
initial changes

Lawful2002 committed:
Merge branch 'web-frag-backend-val' of https://github.com/Lawful2002/oppia into topic-beam-jobs

Lawful2002 committed:
[skip ci]

Lawful2002 committed:
fixed git issues

Lawful2002 committed:
[skip ci] fixed git issues

Lawful2002 committed:
[skip ci] fixed git issues

Lawful2002 committed:
[skip ci] added tests

Lawful2002 committed:
added tests

Lawful2002 committed:
fixed tests

Lawful2002 committed:
fixed failing tests

Lawful2002 committed:
fixed tests

Lawful2002 committed:
fixed tests

Lawful2002 committed:
fixed tests

Lawful2002 committed:
fixed tests

Lawful2002 committed:
fixed tests:

Lawful2002 committed:
fixed tests

Lawful2002 committed:
fixed tests

Lawful2002 committed:
Merge branch 'oppia:develop' into topic-beam-jobs

Lawful2002 commented:
Validation added in PR #15423
